### PR TITLE
Hypothesis support

### PR DESCRIPTION
**Pull request by JCCR**
(Related) https://github.com/readium/readium-shared-js/pull/396

This pull request makes the reader's share URL always show up in the address bar and updates at every location change. This makes use of the browser's [History API](https://developer.mozilla.org/en-US/docs/Web/API/History)
This is needed for Hypothesis so that it could pick up on this URL and use it for sharing purposes as a direct link to a bookmarked location on the current page.

---

Hypothesis has some suggested UI tweaks to our UI that I have implemented as well:

- Providing a margin on the right for the sidebar when it's collapsed:
image

- Shifting the right page turn button when the sidebar is open.
image

### CODE REVIEW

File: src/js/EpubReader.js [hunk 1]
```diff
@@ -1017,6 +1041,19 @@ BookmarkData){
...
+                onLayoutChange: function(state) {
+                    var $rightPageButton = $('#right-page-btn');
+                    $rightPageButton.css('right', state.width);
+                    if (!state.expanded) {
+                        $('nav').css('margin-right', state.width);
+                    }
+                }
+            };
+        };
+
```
Comment (danielweck) [inline src/js/EpubReader.js]: Obviously a bit "hacky". Another dirty implementation trick (albeit slightly less dirty, IMO) would be to place this code in the Hypothes.is plugin:
https://github.com/readium/readium-shared-js/blob/develop/plugins/hypothesis/main.js#L21
See merged PR:
https://github.com/readium/readium-shared-js/pull/396

File: src/js/EpubReader.js [hunk 2]
```diff
@@ -724,9 +724,33 @@ BookmarkData){
...
+        ebookURL = ensureUrlIsRelativeToApp(ebookURL);
+
+        var url = Helpers.buildUrlQueryParameters(undefined, {
+            epub: ebookURL,
+            epubs: " ",
+            embedded: " ",
+            goto: bookmark
+        });
+
+        history.replaceState({}, "", url);
```
Comment (danielweck) [inline src/js/EpubReader.js]: I would like to understand this change better. Why is this required (does this fix a bug?), is this specific to Hypethesi.is integration, or does this offer generic benefits?

Comment (danielweck) [inline src/js/EpubReader.js]: Oh I see, the goto URL parameter is automatically injected! So this would impact the Chrome app as well as the cloud reader. I need to think about the consequences a bit more (concerned about potential regression bugs).

Comment (JCCR) [inline src/js/EpubReader.js]: Yes, that is correct. I would like to see this as a generic benefit, it makes the most sense for the cloud reader only to have this change.

File: src/js/EpubReader.js (hunk 1 above)
Comment (JCCR) [inline src/js/EpubReader.js]: Updated the code in this PR and the plugin with something based on your suggestion

File: src/js/EpubReader.js (hunk 2 above)
Comment (danielweck) [inline src/js/EpubReader.js]: I think the call to `history.replaceState()` should take the following parameters:
`{epub: epub, epubs: epubs},`, `"Readium Viewer"` and `url`, to mimic existing code:
https://github.com/readium/readium-js-viewer/blob/fc41d5be39a8c12d4f142f2c594ad93ba726fde9/src/js/ReadiumViewer.js#L103-L130

Comment (danielweck) [inline src/js/EpubReader.js]: Should there be some sanity checks for the `history` API invocation? Like here:
https://github.com/readium/readium-js-viewer/blob/fc41d5be39a8c12d4f142f2c594ad93ba726fde9/src/js/ReadiumViewer.js#L39-L43
